internal/rbd/encryption.go

```diff
@@ -175,21 +184,63 @@ func (rv *rbdVolume) initKMS(ctx context.Context, volOptions, credentials map[st
 		return nil
 	}
 
-	rv.Encrypted, err = strconv.ParseBool(encrypted)
+	isEncrypted, err := strconv.ParseBool(encrypted)
 	if err != nil {
 		return fmt.Errorf(
 			"invalid value set in 'encrypted': %s (should be \"true\" or \"false\")", encrypted)
-	} else if !rv.Encrypted {
+	} else if !isEncrypted {
 		return nil
 	}
 
-	// deliberately ignore if parsing failed as GetKMS will return default
-	// implementation of kmsID is empty
-	kmsID := volOptions["encryptionKMSID"]
-	rv.KMS, err = util.GetKMS(rv.Owner, kmsID, credentials)
+	err = rv.configureEncryption(volOptions["encryptionKMSID"], credentials)
 	if err != nil {
 		return fmt.Errorf("invalid encryption kms configuration: %w", err)
 	}
 
 	return nil
 }
+
+func (rv *rbdVolume) configureEncryption(kmsID string, credentials map[string]string) error {
+	kms, err := util.GetKMS(rv.Owner, kmsID, credentials)
+	if err != nil {
+		return err
+	}
+
+	rv.encryption, err = util.NewVolumeEncryption(kms)
+
+	// if the KMS can not store the DEK itself, we'll store it in the
```

Comment:
above is bit unclear to me. In which situation or example the KMS cannot store the DEK itself ? Is it referring to the secret based encryption here ?

Reply:
Yes, for the `SecretsKMS` base `SecretsMetadataKMS` that is introduced with this PR. Also it is not wished to stored DEKs in Amazon KMS, as the storing of secrets is more costly than storing the DEK in the metadata of the volume.

Reply:
oh.. ok.. thanks for confirming.